image/image.go

```diff
@@ -153,38 +172,105 @@ func sanitize(s string) string {
 	return strings.Replace(s, "/", "-", -1)
 }
 
-// getSchema1Manifest parses the manifest into a data structure, cleans it up, and returns it.
-// NOTE: The manifest may have been modified in the process; DO NOT reserialize and store the return value
-// if you want to preserve the original manifest; use the blob returned by Manifest() directly.
-// NOTE: It is essential for signature verification that the object is computed from the same manifest which is returned by Manifest().
-func (i *genericImage) getSchema1Manifest() (genericManifest, error) {
-	manblob, err := i.Manifest()
+// compile-time check that manifestSchema2 implements genericManifest
+var _ genericManifest = (*manifestSchema2)(nil)
+
+type manifestSchema2 struct {
+	src               types.ImageSource
+	ConfigDescriptor  descriptor   `json:"config"`
+	LayersDescriptors []descriptor `json:"layers"`
+}
+
+type descriptor struct {
+	MediaType string `json:"mediaType"`
+	Size      int64  `json:"size"`
+	Digest    string `json:"digest"`
+}
+
+func (m *manifestSchema2) String() string {
+	return ""
+}
+
+func (m *manifestSchema2) Layers() []string {
+	blobs := []string{}
+	for _, layer := range m.LayersDescriptors {
+		blobs = append(blobs, layer.Digest)
+	}
+	return blobs
+}
```

Comment:
https://github.com/projectatomic/skopeo/pull/102/files#diff-da07f373046a0a24e67b18699dc15a70R167 does, AFAICS, cause the config to be copied.  And we _need_ that, otherwise the copied image won’t work.

(I’d say that the `genericManifest` is there to be useful for _us_, not to be constrained by the underlying image formats, so what they call things and where they store them should generally be hidden by the `genericManifest` implementations.)

To be more precise, I am thinking:
 - Keep `genericManifest.Layers`, to be used in `ImageInspectInfo` only (for now?); perhaps rename to `LayerDigests` but that is bike-shedding.
 - Add `genericManifest.BlobDigests` = all needed blobs, layers, blobs, anything else.
 - `s/uniqueLayerDigests/uniqueBlobDigests`, calling `BlobDigests` (which also simplifies the semantics, we don't have to keep the disclaimer about layer ordering and whatnot).
 - Replace `types.Image.LayerDigests` with `types.Image.BlobDigests`
 - Modify `skopeo copy` to use `BlobDigests`

There is no guarantee that the manifest+digests of blobs model will work for all possible future image types, but it seems to work fine for v2s1, v2s2 and OCI.